Test/ Slider component interaction test

iyu88 opened the pull request:
## 🤠 개요

- #128 
- Slider 컴포넌트 스토리에 대해서 E2E 테스트를 진행합니다.

## 💫 설명

- Storybook으로 작성한 Slider 컴포넌트의 각 스토리에 E2E 테스트 코드를 작성했습니다. 

- `Default slider`에 대한 테스트 코드를 기준으로 설명드리면,

  - 🐭 마우스 이벤트

  1. Click min value : slider track의 가장 왼쪽 혹은 아래쪽을 클릭했을 때 최솟값으로 설정되는지

  2. Click max value : slider track의 가장 오른쪽 혹은 위쪽을 클릭했을 때 최댓값으로 설정되는지

  - 🎹 키보드 이벤트

  3. Increase with right arrow : ➡️ key를 한번 눌렀을 때, 1 (step) 만큼 증가하는지

  4. Increase with up arrow : ⬆️ key를 한번 눌렀을 때, 1 (step) 만큼 증가하는지

  5. Decrease with left arrow : ⬅️ key를 한번 눌렀을 때, 1 (step) 만큼 감소하는지

  6. Decrease with down arrow : ⬇️ key를 한번 눌렀을 때, 1 (step) 만큼 감소하는지

  7. Increase with page up : `PageUp key`를 눌렀을 때, 10% 증가하는지

  8. Decrease with page down : `PageDown key`를 눌렀을 때, 10% 감소하는지

  9. Set min with home key : `Home key`를 눌렀을 때, 최솟값으로 설정되는지

  10. Set max with end key : `End key`를 눌렀을 때, 최댓값으로 설정되는지

- 그 외의 스토리에 대해서는 필요하다고 판단한 테스트 케이스만 추가했습니다. 

- 추상화하지 않는 것이 테스트를 읽는 것에 더 도움이 될 것이라고 판단하여 일부러 반복되는 부분을 남겨둔 점도 있습니다.

- cypress/support/commands.ts 파일에 `커스텀 커맨드`를 정의했습니다. 
  
  *커스텀 커맨드
  Cypress에서 제공하는 커맨드(메서드)를 조합해서 커스텀할 수 있습니다. 
  Chaining해서 사용할 수 있는 함수를 정의했다고 보시면 될 것 같습니다. 

- cypress/support/index.d.ts 파일에 커스텀 커맨드 및 기타 함수에서 사용할 타입을 정의했습니다.

## 📷 스크린샷 (Optional)

➡️ 58개의 테스트를 작성했습니다.
<img width="534" alt="스크린샷 2023-05-30 22 27 47" src="https://github.com/c-h-w-h/cds/assets/31645195/33719fae-6314-4453-97b6-43b439a8f194">

iyu88 committed:
feat: custom command 추가

iyu88 committed:
test: Slider Click 이벤트 테스트

iyu88 committed:
feat: visitStory command 추가

iyu88 committed:
refactor: 하위 컴포넌트 ID 수정

iyu88 committed:
feat: alias 초기화 함수 선언

iyu88 committed:
feat: UI Paint 확인 함수 추가

iyu88 committed:
test: Arrow Key 테스트 케이스 추가

iyu88 committed:
test: Arrow Key 테스트 케이스 (up, down) 추가

iyu88 committed:
test: Page Up, Down 키보드 이벤트 추가

iyu88 committed:
test: Home, End 테스트 케이스 추가

iyu88 committed:
feat: afterEach에서 shouldBeEqualPercentage 수행

iyu88 committed:
refactor: shouldBeEqualPercentage 수정

함수 내부에서 getComputedStyle를 사용하여 window에서 표시되는 스타일 값으로 비교합니다.

iyu88 committed:
test: min value variant 테스트 케이스 추가

iyu88 committed:
test: max value variant 테스트 케이스 추가

iyu88 committed:
test: With steps 테스트 케이스 추가

iyu88 committed:
test: Start variant 테스트 케이스 추가

iyu88 committed:
test: Vertical 테스트 케이스 추가

prayinforrain commented on cypress/e2e/components/slider.cy.ts:
```diff
@@ -0,0 +1,397 @@
+    });
+
+    it('Increase with right arrow', () => {
+      cy.get("@thumb").trigger('keydown', { keyCode : 39 });
```
키보드 접근성 구현은 많은 컴포넌트에 되어 있으니 이것도 constants같은걸로 정의해서 쓰면 괜찮지 않을까 하는 생각이 들어요 ?ㅁ?
아직 직접 작성해보지 않아서 모르겠지만요 ㅎㅎ..

iyu88 replied:
좋은 의견입니다.
상수 관리를 어떻게 할지 레퍼런스를 찾아볼게요

se030 replied:
이미 constants에 key값 정의되어 있어서 그거 사용하면 컴포넌트 코드랑 통일성 유지 면에서도 좋지 않을까요? 어떤 테스트가 fail했을 때 어떤 부분을 확인해야 하는지 추적하기 편할 것 같아요. keycode는 deprecated property이기도 해서 사용 지양하는게 ..ㅎㅎ!

prayinforrain commented on cypress/e2e/components/slider.cy.ts:
```diff
@@ -0,0 +1,397 @@
+    cy.get('@slider').find(THUMB_ID).as('thumb');
+  }
+
+  const shouldBeEqualPercentage = (dimension: FilledDimension, position: ThumbPosition) => {
```
이 함수들은 `async-await`을 사용하면 안되는 건가요? 아니면 then을 사용한 코드가 더 좋은 이유가 있다던가?
제가 쓰고 싶은 건 아니고 테스트 코드들을 보면 다 이랬던 것 같아서요 ㅋㅋ..

iyu88 replied:
일단 공식 문서에서는 then 문법만 제시하고 있고
별도로 async-await을 적용하고 싶을 경우에는 [cypress-promise](https://github.com/NicholasBoll/cypress-promise#readme)를 적용해야 하는 걸로 파악했어요.
API 요청도 없어서 기본 기능 (일반 함수, alias) 내에서 작성할 수 있을 것 같아서 설치하진 않았어요.

iyu88 replied:
근데 저도 아직 다 파악한건 아니라서 좋은거 찾으시면 알려주세요 ㅎㅎ..

prayinforrain commented on cypress/e2e/components/slider.cy.ts:
```diff
@@ -0,0 +1,397 @@
+
+describe('Slider component test', () => {
+  const initAlias = (storyName: string) => {
+    cy.getStory('slider', storyName).as('slider');
+    cy.get('@slider').find(TRACK_ID).as('track');
+    cy.get('@slider').find(FILLED_ID).as('filled');
+    cy.get('@slider').find(THUMB_ID).as('thumb');
```
`as('name')`으로 선언하면 그 이후의 코드에서 `@name`이라는 alias로 참조할 수 있는 거 맞나요?

iyu88 replied:
네 정확합니다 👍👍👍

prayinforrain reviewed (approved):
고생하셨습니다~ 너무 신기해요

근데 제 환경에서는 테스트가 간헐적으로 실패해요. 컴퓨터 성능에 따라 좀 갈리는 것 같긴 한데(제가 작업을 HDD에 하고 있어서 ㅎㅎ..) 모두 같은 지점에서 실패하는걸로 보아 스토리를 불러오는 게 오래 걸려서 타임아웃으로 alias 정의를 못하나봐요. 사실 제 컴퓨터에서 안되는건 상관이 없긴 한데 혹시 해결할 방법이 있나 찾아봐야겠어요.
```
Error: AssertionError: Timed out retrying after 4000ms: Expected to find element: `#story--components-slider--start-from-end`, but never found it. Queried from:

              > <body.sb-show-main.sb-main-fullscreen>

Because this error occurred during a `before each` hook we are skipping the remaining tests in the current suite: `Start From End slider`
    at Context.eval (webpack:///./cypress/support/commands.ts:7:0)
```
![image](https://github.com/c-h-w-h/cds/assets/46566891/ae94ae1d-beda-4ffd-9981-0b938fc9799e)

![image](https://github.com/c-h-w-h/cds/assets/46566891/19aaa6ad-8be3-4a7e-9842-6ab87daafb5a)

iyu88 commented:
Timed out 관련 문제가 있네요! 알아보겠습니다 ㅎㅎ

leesunmin1231 reviewed (commented):
제 컴퓨터에서도 우재님이 말씀하신 것처럼 한번씩 테스트 통과가 안돼요.. 매번 통과 안되는 지점도 바뀌네요. 
오류 메시지로 추측해봤을 때는 비동기로 동작하는 로직이 순서 보장이 안돼서 테스트 통과가 되었다 안되었다 하는거 같아요
<img width="1069" alt="image" src="https://github.com/c-h-w-h/cds/assets/81913106/2346bdfb-f871-4e22-848c-a1423c3c0b5d">

leesunmin1231 reviewed (commented):
제 컴퓨터에서도 우재님이 말씀하신 것처럼 한번씩 테스트 통과가 안돼요.. 매번 통과 안되는 지점도 바뀌네요. 
오류 메시지로 추측해봤을 때는 비동기로 동작하는 로직이 순서 보장이 안돼서 테스트 통과가 되었다 안되었다 하는거 같아요
<img width="1069" alt="image" src="https://github.com/c-h-w-h/cds/assets/81913106/2346bdfb-f871-4e22-848c-a1423c3c0b5d">

se030 reviewed (commented):
실행해보니 다른 분들이랑 같은 이슈가 발생하기는 하는데, 선민님 말씀처럼 비동기 로직들의 순서가 보장되지 않아서 그런 것 같아요. 저도 공부해보고 뭔가 더 알게되면 공유할게요!

iyu88 committed:
feat: visitStory 에서 개별 스토리 방문 URL 수정

스토리 Docs에서  iframe 내부에 있는 DOM 요소에 접근하는 방식으로 visitStory가 작성되어 있던 점을 수정합니다.
개별 스토리에 대한 iframe 내부로 바로 접근할 수 있도록 URL 주소를 수정합니다.
이로 인해서 getStory 커맨드는 더 이상 사용하지 않게 되어 삭제합니다.
더불어 initAlias 함수에서 스토리의 최상위 부모 요소인 #root로부터 하위 컴포넌트를 찾도록 변경합니다.
아울러 공통적으로 적용되는 beforeEach는 삭제합니다.
